forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceBindingCallback.kt

```diff
@@ -168,41 +202,21 @@ open class DeviceBindingCallback: AbstractCallback {
      * @param status  DeviceBindingStatus(timeout,Abort, unsupported)
      * @param listener The Listener to listen for the result
      */
-    protected open fun handleException(status: DeviceBindingStatus,
+    protected open fun handleException(status: DeviceBindingStatus<Any>,
+                                       e: Exception?,
                                        listener: FRListener<Void>) {
 
         setClientError(status.clientError)
-        Logger.error(tag, status.message, status.errorCode)
-        Listener.onException(
-            listener,
-            DeviceBindingException(status)
-        )
-    }
-
-    /**
-     * create the interface for the Authentication type(Biometric, Biometric_Fallback, none)
-     */
-    protected open fun getDeviceBindAuthenticator(): DeviceAuthenticator {
-        return AuthenticatorFactory.getType(userId, deviceBindingAuthenticationType, title, subtitle, description)
+        Logger.error(tag, e, status.message)
+        Listener.onException(listener, DeviceBindingException(status, e))
     }
 
-    /**
-     * Get Expiration date for the signed token, claim "exp" will be set to the JWS.
-     *
-     * @return The expiration date
-     */
-    protected open fun getExpiration(): Date {
-        val date = Calendar.getInstance();
-        date.add(Calendar.SECOND, timeout ?: 60)
-        return date.time;
-    }
 }
 
 /**
  * convert authentication string received from server to authentication enum
  */
 enum class DeviceBindingAuthenticationType constructor(val serializedValue: String?) {
-    BIOMETRIC_ONLY("BIOMETRIC_ONLY"),
-    BIOMETRIC_ALLOW_FALLBACK("BIOMETRIC_ALLOW_FALLBACK"),
-    NONE("NONE");
+    BIOMETRIC_ONLY("BIOMETRIC_ONLY"), BIOMETRIC_ALLOW_FALLBACK("BIOMETRIC_ALLOW_FALLBACK"), NONE("NONE"), APPLICATION_PIN(
+        "APPLICATION_PIN");
```

Comment:
we can extend this class and it make sense to introduce a extension for this. we can kill the AuthenticatorFactory Example:

 fun DeviceBindingAuthenticationType.getType(context: Context): DeviceAuthenticator {
    return when (this) {
        DeviceBindingAuthenticationType.BIOMETRIC_ONLY -> BiometricOnly()
        DeviceBindingAuthenticationType.APPLICATION_PIN -> ApplicationPinDeviceAuthenticator(context)
        DeviceBindingAuthenticationType.BIOMETRIC_ALLOW_FALLBACK -> BiometricAndDeviceCredential()
        else -> None()
    }
}

Reply:
optional: But we can do more on this here 

`fun DeviceAuthenticator.intialize() {
     when (this) {
      is BiometricOnly -> {
          this.setKeyAware()
      }
      is ApplicationPinDeviceAuthenticator ->    {

      }
    }
}
`

Reply:
Updated